src/Nethermind/Nethermind.Mev/Source/BundlePool.cs

```diff
@@ -61,238 +85,246 @@ public BundlePool(IBlockTree blockTree, IBundleSimulator simulator, IBlockFinali
+                Task<SimulatedMevBundle> simulateTask = _simulator.Simulate(bundle, parent, cancellationTokenSource.Token);
+                simulateTask.ContinueWith(TryRemoveFailedSimulatedBundle, cancellationTokenSource.Token);
+                return context = new(simulateTask, cancellationTokenSource);
+            }
+            
+            ConcurrentDictionary<(MevBundle, Keccak), SimulatedMevBundleContext> AddContext(ConcurrentDictionary<(MevBundle, Keccak), SimulatedMevBundleContext> d)
             {
-                context.Task = _simulator.Simulate(bundle, parent, context.CancellationTokenSource.Token);
+                d.AddOrUpdate(key, 
+                    _ => CreateContext(), 
+                    (_, c) => c);
+                return d;
             }
+            
+            _simulatedBundles.AddOrUpdate(parent.Number, 
+                    _ =>
+                    {
+                        ConcurrentDictionary<(MevBundle, Keccak), SimulatedMevBundleContext> d = new();
+                        return AddContext(d);
+                    },
+                    (_, d) => AddContext(d));
+
+            return context;
+        }
 
-            ConcurrentBag<Keccak> blocksBag;
-            lock (_bundles)
+        private void TryRemoveFailedSimulatedBundle(Task<SimulatedMevBundle> simulateTask)
+        {
+            if (simulateTask.IsCompletedSuccessfully)
             {
-                blocksBag = _bundles[bundle];
+                SimulatedMevBundle simulatedMevBundle = simulateTask.Result;
+                if (!simulatedMevBundle.Success)
+                {
+                    _bundles.TryRemove(simulatedMevBundle.Bundle);
+                    RemoveSimulation(simulatedMevBundle.Bundle);
+                }
             }
-            blocksBag.Add(parentHash);
         }
-        
-        private void OnNewSuggestedBlock(object? sender, BlockEventArgs e)
+
+        private void OnNewBlock(object? sender, BlockEventArgs e)
         {
             long blockNumber = e.Block!.Number;
-            if (_finalizationManager?.IsFinalized(blockNumber) != true)
+            RemoveBundlesUpToBlock(blockNumber);
+
+            Task t = Task.Run(() => //does this need to be in a task?
             {
-                Task.Run(() =>
+                UInt256 timestamp = _timestamper.UnixTime.Seconds;
+                IEnumerable<MevBundle> bundles = GetBundles(e.Block.Number + 1, UInt256.MaxValue, timestamp);
+                foreach (MevBundle bundle in bundles)
                 {
-                    IEnumerable<MevBundle> bundles = GetBundles(e.Block.Number + 1, UInt256.MaxValue, UInt256.Zero);
-                    foreach (MevBundle bundle in bundles)
-                    {
-                        SimulateBundle(bundle, e.Block.Header);
-                    }
-                });
+                    SimulateBundle(bundle, e.Block.Header);
+                }
+            });
+            t.Wait();
```

Comment:
this is  wrong, this is meant to be async, you shouldn't be calling Task.Wait almost never.